# mHapSc
mHapSc is a tool kit for analysis of single cell DNA methylation haplotypes.
It has 7 sub-commands: convert, tanghulu, R2, MHBDiscovery, hemiM, track and flinkage.
The source code of mHapSc is hosted in [github](https://github.yoyoong/mHapSc). It is written in java, and should run in Windows, Linux and macOS with java1.8 installed.

## Prerequisites
The mHapSc requires the following dependencies:
* java jdk 1.8.0

## Installation
Push the source and run the jar:
```Shell
git clone https://github.yoyoong/mHapSc
cd mHapSc/target
java -jar mHapSc-1.0-jar-with-dependencies.jar [convert/merge/tanghulu/mHapView/MHBDiscovery/stat/genomeWide ...]
```

## convert
Convert Single-cell methylation sequencing file or nanopore methylation sequencing file to mHap format file. 
Single-cell methylation sequencing file was introduced in [scMethBank-Documentation](https://ngdc.cncb.ac.cn/methbank/scm/documentation#data).
Nanopore methylation sequencing file was introduced in [Simpson JT et al., 2017](https://pubmed.ncbi.nlm.nih.gov/28218898/)
```Shell
usage: java -jar mHapSc-1.0-jar-with-dependencies.jar convert --inputPath <in.bed/tsv.gz> --cpgPath <CpG.gz> [--region chr:start-end] --outputDir <outputDir> --tag <tag> [--nanopolish]
Options:
 -inputPath,--inputPath <args>   input file, Single-cell methylation sequencing file is bed.gz format, Nanopore methylation sequencin is tsv.gz format
 -cpgPath,--cpgPath <args>       genomic CpG file, gz format and indexed
 -region,--region <args>         one region, in the format of chr:start-end
 -outputDir,--outputDir <args>   output directory, created in advance
 -tag,--tag <args>               prefix of the output file(s)
 -nanopolish,--nanopolish        indicates whether input file is nanopore methylation sequencing file
 ```
Example of usage:
```Shell
java -jar mHapSc-1.0-jar-with-dependencies.jar convert -inputPath methylation-log.tsv.gz -cpgPath hg38_CpG.gz -outputDir outputDir -tag nanopolish_test -nanopolish
```

## tanghulu
Visualize the methylation status of the reads that cover a given region.
Tanghulu plot was originally introduced in [CGmapTools](https://cgmaptools.github.io/).
```Shell
usage: java -jar mHapSc-1.0-jar-with-dependencies.jar tanghulu --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--bcFile <bcFile>] --region chr:start-end --outputDir <outputDir> --tag <tag> [--outcut outcut]
Options:
 -mhapPath,--mhapPath <args>     input file,mhap.gz format,generated by mHapTools and indexed
 -cpgPath,--cpgPath <args>       genomic CpG file, gz format and indexed
 -bcFile,--bcFile <args>         barcode file
 -region,--region <args>         one region, in the format of chr:start-end
 -outputDir,--outputDir <args>   output directory, created in advance
 -tag,--tag <args>               prefix of the output file(s)
 -outcut,--outcut <args>         outcut
 ```
Example of usage:
```Shell
java -jar mHapSc-1.0-jar-with-dependencies.jar tanghulu -mhapPath CRC_hg19.mhap.gz -cpgPath hg19_CpG.gz -region chr1:3229374-3230473 -outputDir outputDir -tag CRC_hg19 -bcFile CRC_LN.txt
```

## R2
R2
```Shell
usage: java -jar mHapSc-1.0-jar-with-dependencies.jar R2 --mhapPath <mhap.gz> --cpgPath <CpG.gz> --region chr:start-end [--bcFile <bcFile>] --outputDir <outputDir> --tag <tag> [--mhapView] [--strand strand] [--longrange] [--bedFile bedFile]
Options:
 -mhapPath,--mhapPath <args>     input file,mhap.gz format,generated by mHapTools and indexed
 -cpgPath,--cpgPath <args>       genomic CpG file, gz format and indexed
 -bcFile,--bcFile <args>         barcode file
 -region,--region <args>         one region, in the format of chr:start-end
 -outputDir,--outputDir <args>   output directory, created in advance
 -tag,--tag <args>               prefix of the output file(s) [out]
 -mHapView,--mHapView            mHapView flag [false]
 -strand,--strand <args>         plus,minus,both [both]
 -longrange,--longrange          longrange flag [false]
 -bedFile,--bedFile <args>       a bed file
 ```
Example of usage:
```Shell
java -jar mHapSc-1.0-jar-with-dependencies.jar R2 -mhapPath CRC_hg19.mhap.gz -cpgPath hg19_CpG.gz -region chr1:3229375-3230473 -bcFile CRC_LN.txt -outputDir outputDir -tag CRC_hg19 -longrange -mHapView -bedFile CRC_sc_bulk.bed 
```
## MHBDiscovery
DNA methylation of adjacent CpG sites can be co-methylated and forms methylation haplotype blocks (MHBs).
It was original introduced in [S. Guo et al., 2017](https://pubmed.ncbi.nlm.nih.gov/28263317/).
```Shell
usage: java -jar mHapSc-1.0-jar-with-dependencies.jar MHBDiscovery --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--bcFile <bcFile>] [--region chr:start-end] [--bedPath <in.bed>] -outputDir outputDir --tag <tag> [--window window] [--r2 r2] [--pvalue pvalue] [--qc]
Options:
 -mhapPath,--mhapPath <args>      input file,mhap.gz format,generated by mHapTools and indexed
 -cpgPath,--cpgPath <args>        genomic CpG file, gz format and indexed
 -bcFile,--bcFile <args>          barcode file
 -region,--region <args>          one region, in the format of chr:start-end
 -bedPath,--bedPath <args>        input BED file
 -outputDir,--outputDir <args>    output directory, created in advance
 -tag,--tag <args>                prefix of the output file(s) [MHBDiscovery.out]
 -window,--window <args>          Size of core window
 -r2,--r2 <args>                  R square cutoff
 -pvalue,--pvalue <args>          P value cutoff
 -qc,--qc                         indicates whether output matrics for QC [false]
 ```
Example of usage:
```Shell
java -jar mHapSc-1.0-jar-with-dependencies.jar MHBDiscovery -mhapPath CRC_hg19.mhap.gz -region chr1:3229375-3230473 -cpgPath hg19_CpG.gz -window 5 -r2 0.5 -pvalue 0.05 -outputDir outputDir -tag CRC_hg19_MHB -qc 
```

## hemiM
hemiM
```Shell
usage: java -jar mHapSc-1.0-jar-with-dependencies.jar hemiM --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--bcFile <bcFile>] [--region chr:start-end] [--bedFile <in.bed>] -outputDir outputDir --tag <tag>
Options:
 -mhapPath,--mhapPath <args>      input file,mhap.gz format,generated by mHapTools and indexed
 -cpgPath,--cpgPath <args>        genomic CpG file, gz format and indexed
 -bcFile,--bcFile <args>          barcode file
 -region,--region <args>          one region, in the format of chr:start-end
 -bedFile,--bedFile <args>        input BED file
 -outputDir,--outputDir <args>    output directory, created in advance
 -tag,--tag <args>                prefix of the output file(s) [hemiM.out]
 ```
Example of usage:
```Shell
java -jar mHapSc-1.0-jar-with-dependencies.jar hemiM -mhapPath CRC_hg19.mhap.gz -cpgPath hg19_CpG.gz -bedFile CRC_MHB_non_NC.bed -bcFile CRC_LN.txt -outputDir outputDir -tag test
```

## track
Calcuate mHap-level metrics for all reads that cover each individual CpG sites in a genome-wide manner and save the results as bedGraph files.
```Shell
usage: java -jar mHapSc-1.0-jar-with-dependencies.jar track --metrics <metrics> --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--bcFile <bcFile>] [--region chr:start-end] [--bedFile <in.bed>] --outputDir <outputDir> --tag <tag>
Options:
 -metrics,--metrics <args>       mHap-level metrics,including MM and R2
 -mhapPath,--mhapPath <args>     input file,mhap.gz format,generated by mHapTools and indexed
 -cpgPath,--cpgPath <args>       genomic CpG file, gz format and indexed
 -bcFile,--bcFile <args>         barcode file
 -region,--region <args>         one region, in the format of chr:start-end
 -bedFile,--bedFile <args>       input BED file
 -outputDir,--outputDir <args>   output directory, created in advance
 -tag,--tag <args>               prefix of the output file(s) [track.out]
 ```
Example of usage:
```Shell
java -jar mHapSc-1.0-jar-with-dependencies.jar track -mhapPath CRC_hg19.mhap.gz -cpgPath hg19_CpG.gz -region chr1:3229375-13230473 -outputDir outputDir -tag test outputDir -tag test -bcFile CRC_LN.txt -metrics MM R2 
```

## flinkage
flinkage
```Shell
usage: java -jar mHapSc-1.0-jar-with-dependencies.jar flinkage --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--bcFile <bcFile>] [--region1 chr:start-end] [--region2 chr:start-end] [--bedFile <in.bed>] --outputDir <outputDir> --tag <tag> [--limit limit]
Options:
 -metrics,--metrics <args>       mHap-level metrics,including MM and R2
 -mhapPath,--mhapPath <args>     input file,mhap.gz format,generated by mHapTools and indexed
 -cpgPath,--cpgPath <args>       genomic CpG file, gz format and indexed
 -bcFile,--bcFile <args>        barcode file
 -region1,--region <args>        the first region, in the format of chr:start-end
 -region2,--region <args>        the second region, in the format of chr:start-end
 -bedFile,--bedFile <args>       input BED file
 -outputDir,--outputDir <args>   output directory, created in advance
 -tag,--tag <args>               prefix of the output file(s) [flinkage.out]
 -limit,--limit <args>           the max length to calculate [20000000]
 ```
Example of usage:
```Shell
java -jar mHapSc-1.0-jar-with-dependencies.jar flinkage -mhapPath CRC_hg19.mhap.gz -cpgPath hg19_CpG.gz -region1 chr1:949817-949850 -region2 chr1:969429-969458 -bcFile CRC_LN.txt -outputDir outputDir -tag flinkage -limit 20000000
```
